### PR TITLE
Add import from git workflow using webview UI

### LINKED ISSUE #2538: Add Import from Git workflow using webview UI
Screenshot 2022-09-15 at 5 53 38 PM

The user provides the git repo location (given it is not added to the local workspace yet). As a user, I need a recommendation on how to deploy this application using devfile.

The extension then analyses the git repo and recommends the following:
- Based on the git repo files and packages, it recommends the `devfile.yaml` based on the analysis.
- If the user accepts the recommendation, it can proceed to create the component. Else the user can provide their own custom `devfile.yaml`.
- The extension then creates a component using the `devfile.yaml` and adds it to the Components View
- The user then does the `Start Dev` command on the created component.

This is useful to the user to directly start from the git repo and then the extension does the rest i.e clone, create and add to the component's view. Thus combining multiple manual steps into 1 step, improves the workflow to start on OpenShift using your git repo URL only.

### PR DESCRIPTION
**Pull request by msivasubramaniaan**
Signed-off-by: msivasubramaniaan [msivasub@redhat.com](mailto:msivasub@redhat.com)

Fix #2538

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -477,6 +487,15 @@
 					"dark": "images/title/dark/workspace-new.svg"
 				}
 			},
+			{
+				"command": "openshift.component.importFromGit",
+				"title": "Import from Git repository",
+				"category": "OpenShift",
+				"icon": {
+					"light": "images/title/light/workspace-new.svg",
```
Comment (mohitsuman) [inline package.json:495]: Add git icon to reflect import from git command.

**Comment by mohitsuman** [comment]
1. Add a Button `Import from Git` in the Application Explorer section. This will directly open the webview.
2. Once the gitrepo is validated and we change the gitrepo (provide a different URL), the Application name and name text box is not updated.

@msivasubramaniaan ^^

File: src/webview/git-import/app/cardItem.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                <div className={style.cardBody}>
+                    {
+                        yamlDoc.metadata.version && (
+                            <Typography variant='caption'>
+                                Version: {yamlDoc.metadata.version}<br />
+                            </Typography>
+                        )
+                    }
+                    <Typography variant='caption'>
+                        Project Type: {yamlDoc.metadata.projectType}<br />
```
Comment (mohitsuman) [inline src/webview/git-import/app/cardItem.tsx:46]: Let's make sure the Type is `NodeJS` and NOT `nodejs`.

File: src/webview/git-import/app/cardItem.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                    </Typography>
+                    <Typography variant='caption'>
+                        Language: {yamlDoc.metadata.language}<br />
```
Comment (mohitsuman) [inline src/webview/git-import/app/cardItem.tsx:49]: Similarly here to verify the name as NodeJS.

File: src/webview/git-import/app/gitImport.tsx [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                                    />
+                                </this.AccordionDetails>
+                            </this.Accordion>
+                            {yamlDoc &&
+                                <CardItem key='cardKey' yamlDoc={yamlDoc} />
+                            }
+                            {applicationName && name &&
+                                <>
+                                    <div className='title sub'>
+                                        <Typography variant='h6' className='highlight'>General</Typography>
```
Comment (mohitsuman) [inline src/webview/git-import/app/gitImport.tsx:219]: We do not need General in the UI. Let's remove it.

File: src/webview/git-import/app/gitImport.tsx [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        padding: theme.spacing(2),
+        borderTop: '1px solid rgba(0, 0, 0, .125)',
+    }));
+
+    render(): React.ReactNode {
+        const { gitURLValid, accordionOpen, applicationName, name, yamlDoc } = this.state;
+        return (
+            <div className='mainContainer margin' >
+                <div className='title'>
+                    <Typography variant='h4' className='highlight'>Import from Git</Typography>
```
[the remark this answers is not in the document]
Comment (mohitsuman) [reply src/webview/git-import/app/gitImport.tsx:128]: Make the heading as `h6`.

File: src/webview/git-import/app/gitImport.tsx [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+                </div>
+                <div className='formContainer'>
+                    <div className='form'>
+                        <InputLabel required htmlFor='bootstrap-input'
+                            style={{
+                                color: 'var(--vscode-settings-textInputForeground)'
```
Comment (mohitsuman) [inline src/webview/git-import/app/gitImport.tsx:134]: Make color as `#fe5142`

File: src/webview/git-import/app/cardItem.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            >
+                <div className={style.cardHeader}>
+                    <div className={style.cardHeaderDisplay}>
+                        <img
+                            src={yamlDoc.metadata.icon}
+                            className={style.cardImage} />
+                    </div>
+                </div>
+                <div style={{ margin: '1.5rem' }}>
+                    <Typography variant='body1'>{yamlDoc.metadata.name}</Typography>
```
Comment (mohitsuman) [inline src/webview/git-import/app/cardItem.tsx:35]: Here also we need to keep it full name. Let's see if we can map `metadata.name` to use fullName.

**Review by mohitsuman — CHANGES_REQUESTED** [review]
Let's get these changes done.

File: src/webview/git-import/app/gitImport.tsx [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+                                </>
+                            }
+                            <div style={{ margin: '2rem 0rem', position: 'fixed' }}>
+                                <Button variant='contained'
+                                    disabled={gitURLValid.showError}
+                                    component='span' className='buttonStyle'>
+                                    Create
+                                </Button>
+                                <Button variant='contained' className='buttonStyle' style={{ marginLeft: '1rem' }}>
+                                    Cancel
```
Comment (mohitsuman) [inline src/webview/git-import/app/gitImport.tsx:267]: let's have the Cancel Button as Outlined parameter as we have in MUI.

**Comment by mohitsuman** [comment]
The Cancel button does not work in Light Theme.

Screenshot 2022-10-22 at 5 00 53 AM

**Review by mohitsuman — APPROVED** [review]

**Review by gorkem — CHANGES_REQUESTED** [review]
This use case needs to be discussed further

**Comment by mohitsuman** [comment]
@gorkem As discussed, can you please approve the PR and we merge it to main.

**Review by mohitsuman — APPROVED** [review]
LGTM

**Review by gorkem — CHANGES_REQUESTED** [review]
rename `src/git` to `src/git-import`

Most of the `/utils` package is about detecting projects. Can we use Alizer instead of reimplementing?

**Comment by dgolovin** [comment]
I get rate limit error right away on first try.
image

File: src/git/services/git-service.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+ *  Copyright (c) Red Hat, Inc. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE file in the project root for license information.
+ *-----------------------------------------------------------------------------------------------*/
+import { GitSource, GitProvider, SecretType as GitSecretType } from '../types';
+import { BaseService } from './base-service';
+import { BitbucketService } from './bitbucket-service';
+import { GithubService } from './github-service';
+import { GitlabService } from './gitlab-service';
+
+export enum SecretType {
```
[the remark this answers is not in the document]
Comment (dgolovin) [reply src/git/services/git-service.ts]: Renamed enum so there is not shadowing, but VSCode still complains
image

**Comment by dgolovin** [comment]
@gorkem renamed git to git-import

**Comment by mohitsuman** [comment]
> rename `src/git` to `src/git-import`
> 
> Most of the `/utils` package is about detecting projects. Can we use Alizer instead of reimplementing?

For the next iteration, we would be looking to implement with alizer. As alizer only works with local workspace and not remote URL, we need to fix it on alizer and then support it on the extension side. I will keep an issue for the next release. For the 1.0 release, we should be good to implement the current project detection strategy.

**Review by mohitsuman — APPROVED** [review]

**Review by dgolovin — APPROVED** [review]

**Pull request merged**